kolibri/core/content/utils/channel_import.py

```diff
@@ -652,73 +702,40 @@ def can_use_sqlite_attach_method(self, model, table_mapper):
-        # that is already limiting the number of elements that will be passed into this.
-        existing_values = {
-            v[pk_column.name]: v
-            for v in self.destination.execute(
-                select([DestinationTable]).where(
-                    pk_column.in_(d[pk_column.name] for d in data)
-                )
-            ).fetchall()
-        }
-        data_to_return = []
-        if do_not_overwrite:
-            for d in data:
-                if d[pk_column.name] not in existing_values:
-                    for column_name, column in columns:
-                        value = d.get(column_name)
-                        if value is None:
-                            value = self.get_and_set_column_default(column)
-                        d[column_name] = value
-                    data_to_return.append(d)
-        else:
-            for d in data:
-                pk = d[pk_column.name]
-                if pk in existing_values:
-                    value = dict(existing_values[pk])
-                    value.update(d)
-                else:
-                    value = d
-                for column_name, column in columns:
-                    if column_name not in value or value[column_name] is None:
-                        value[column_name] = self.get_and_set_column_default(column)
-                data_to_return.append(value)
-        return data_to_return
-
     def check_and_delete_existing_channel(self):
         ChannelMetadataTable = self.destination.get_table(ChannelMetadata)
         existing_channel = self.destination.execute(
-            select([ChannelMetadataTable]).where(
+            select(ChannelMetadataTable).where(
                 ChannelMetadataTable.c.id == self.channel_id
             )
         ).fetchone()
 
         if existing_channel:
-
-            if existing_channel["version"] < self.channel_version:
+            if self.partial:
+                if existing_channel["version"] != self.channel_version:
+                    # We have previously loaded this channel, with a different version to the metadata we are trying to insert
+                    logger.warning(
+                        (
+                            "Version {channel_version} of channel {channel_id} already exists in database; cancelling partial import of "
+                            + "version {new_channel_version}"
+                        ).format(
+                            channel_version=existing_channel["version"],
+                            channel_id=self.channel_id,
+                            new_channel_version=self.channel_version,
+                        )
+                    )
+                    return False
+                return True
+            elif existing_channel["version"] < self.channel_version:
                 # We have an older version of this channel, so let's clean out the old stuff first
```

Comment:
Seems this should inherently and correctly handle the case where the previous import was a partial import?

Reply:
As it just deletes everything, it will handle the case when the previous import was partial or when it was complete. We add the check for a current partial import, because we can only do that in the case where the currently installed version of the channel is the same, otherwise we run the risk of mismatched/colliding tree metadata.